rpc/src/client/transport/router.rs

```diff
@@ -0,0 +1,182 @@
+    /// the event's.
+    pub fn publish(&mut self, ev: &Event) -> PublishResult {
+        let subs_for_query = match self.subscriptions.get_mut(&ev.query) {
+            Some(s) => s,
+            None => return PublishResult::NoSubscribers,
+        };
+        // We assume here that any failure to publish an event is an indication
+        // that the receiver end of the channel has been dropped, which allows
+        // us to safely stop tracking the subscription.
+        let mut disconnected = HashSet::new();
+        for (id, event_tx) in subs_for_query {
+            if let Err(e) = event_tx.send(Ok(ev.clone())) {
+                disconnected.insert(id.clone());
+                debug!(
+                    "Automatically disconnecting subscription with ID {} for query \"{}\" due to failure to publish to it: {}",
+                    id, ev.query, e
+                );
+            }
+        }
+        // Obtain a mutable reference because the previous reference was
+        // consumed in the above for loop. We should panic if there are no
+        // longer any subscriptions for this query.
+        let subs_for_query = self.subscriptions.get_mut(&ev.query).unwrap();
+        for id in disconnected {
+            subs_for_query.remove(&id);
+        }
+        if subs_for_query.is_empty() {
+            PublishResult::AllDisconnected
+        } else {
+            PublishResult::Success
+        }
+    }
+
+    /// Immediately add a new subscription to the router without waiting for
+    /// confirmation.
+    pub fn add(&mut self, id: impl ToString, query: impl ToString, tx: SubscriptionTx) {
+        let query = query.to_string();
+        let subs_for_query = match self.subscriptions.get_mut(&query) {
+            Some(s) => s,
+            None => {
+                self.subscriptions.insert(query.clone(), HashMap::new());
+                self.subscriptions.get_mut(&query).unwrap()
+            }
+        };
+        subs_for_query.insert(id.to_string(), tx);
+    }
+
+    /// Removes all the subscriptions relating to the given query.
+    pub fn remove_by_query(&mut self, query: impl ToString) -> usize {
+        self.subscriptions
+            .remove(&query.to_string())
+            .map(|subs_for_query| subs_for_query.len())
+            .unwrap_or(0)
+    }
+}
+
+#[cfg(feature = "websocket-client")]
+impl SubscriptionRouter {
+    /// Returns the number of active subscriptions for the given query.
+    pub fn num_subscriptions_for_query(&self, query: impl ToString) -> usize {
+        self.subscriptions
+            .get(&query.to_string())
+            .map(|subs_for_query| subs_for_query.len())
+            .unwrap_or(0)
+    }
+}
+
+impl Default for SubscriptionRouter {
+    fn default() -> Self {
+        Self {
+            subscriptions: HashMap::new(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum PublishResult {
+    Success,
+    NoSubscribers,
+    AllDisconnected,
```

Comment:
what is the intuition behind this type of result?

Reply:
It was one way of passing some additional information back to the `WebSocketClient` after it attempts to publish an event.

Since the `WebSocketClient` only sends an unsubscribe request once all subscribers for a particular query have dropped, I needed to be able to distinguish between the case where there were no subscribers to begin with (i.e. we somehow received a spurious event without having a subscription) and the case where we had subscribers, but since we published last to them they dropped their receiving ends. Both would publish the event to zero subscribers.

In the former case, we would consider this to be unexpected behaviour on the part of the node (I don't do anything about it right now though), and in the latter case it'd be a signal to send a single unsubscribe request.